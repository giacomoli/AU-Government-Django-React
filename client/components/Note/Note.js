import React from 'react'
import { graphql, compose } from 'react-apollo'
import gql from 'graphql-tag'
import moment from 'moment'
import { reduxForm, Field, change } from 'redux-form'

import { getMatter } from 'queries'
import CustomDatePicker from 'components/DatePicker'
// import { noteFragment } from 'fragments'
import { renderTextarea } from 'utils'

class Note extends React.PureComponent {
  constructor(props) {
    super(props)

    this.state = {
      editMode: false,
      date: moment(props.note.dateTime),
    }
  }
  handleOnSave = async (values) => {
    const noteId = this.props.note.id
    const { user } = this.props
    const { data } = await this.props.updateNote({
      variables: { ...values, noteId: this.props.note.id, userId: user.id },
      update: (store, { data: { updateNote } }) => {
        const storeData = store.readQuery({
          query: getMatter,
          variables: { id: this.props.matterId },
        })
        const index = storeData.matter.notes.findIndex(note => note.id === noteId)

        if (index > -1) {
          storeData.matter.notes[index] = updateNote.note
        }
        store.writeQuery({
          query: getMatter,
          variables: { id: this.props.matterId },
          data: storeData,
        })
      },
      optimisticResponse: {
        __typename: 'Mutation',
        updateNote: {
          errors: [],
          __typename: 'UpdateNoteMutation',
          note: {
            __typename: 'NoteType',
            id: noteId,
            dateTime: values.dateTime,
            text: values.text,
            user: {
              id: user.id,
              fullName: user.fullName,
              __typename: 'UserType',
            },
          },
        },
      },
    })

    if (data.updateNote.errors.length > 0) {
      console.log(data.updateNote.errors)
    } else {
      this.setState({ editMode: false })
    }
  }
  handleDateChange = (date) => {
    this.props.dispatch(change(this.props.form, 'dateTime', date.format()))
    this.setState({ date })
  }
  enterEditMote = () => {
    this.props.dispatch(change(this.props.form, 'text', this.props.note.text))
    this.props.dispatch(change(this.props.form, 'dateTime', this.props.note.dateTime))
    this.setState({ editMode: true })
  }
  leaveEditMode = () => this.setState({ editMode: false })
  renderDateTime = () => {
    if (this.state.editMode) {
      return (
        <div>
          <Field component="input" type="hidden" name="dateTime" />
          <CustomDatePicker selected={this.state.date} onChange={this.handleDateChange} />
        </div>
      )
    }

    return (
      <div onClick={this.enterEditMote} tabIndex={0} role="button" className="col">
        {moment(this.props.note.dateTime).format('MMM Do YYYY')}
      </div>
    )
  }
  renderNoteText = () => {
    if (this.state.editMode) {
      return [
        <div key={1} className="col">
          <Field component={renderTextarea} formGroup={false} name="text" />
        </div>,
        <div key={2} className="col-sm-auto align-self-start btn-group">
          <button
            type="button"
            onClick={this.props.handleSubmit(this.handleOnSave)}
            className="btn btn-success"
          >
            <i className="fa fa-check" />
          </button>
          <button type="button" onClick={this.leaveEditMode} className="btn btn-success">
            <i className="fa fa-times" />
          </button>
        </div>,
      ]
    }

    return (
      <div onClick={this.enterEditMote} tabIndex={0} role="button" className="col">
        {this.props.note.text}
      </div>
    )
  }
  render() {
    return (
      <tr>
        <td style={{ width: 200, textAlign: 'center' }}>{this.renderDateTime()}</td>
        <td>
          <div className="row">{this.renderNoteText()}</div>
        </td>
        <td>{this.props.note.user && this.props.note.user.fullName}</td>
        <style jsx>{`
          tr:hover {
            cursor: pointer;
          }
        `}</style>
      </tr>
    )
  }
}

const updateNote = gql`
  mutation updateNote($noteId: ID!, $text: String!, $dateTime: String!, $userId: ID!) {
    updateNote(noteId: $noteId, text: $text, dateTime: $dateTime, userId: $userId) {
      errors
      note {
        id
        text
        dateTime
        user {
          id
          fullName
        }
      }
    }
  }
`

export default reduxForm({})(
  compose(graphql(updateNote, { name: 'updateNote' }))(
    Note,
  ),
)
