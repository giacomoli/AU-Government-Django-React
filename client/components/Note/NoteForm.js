import React from 'react'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import { reduxForm, Field, reset } from 'redux-form'
import { connect } from 'react-redux'
import moment from 'moment'

import { noteFragment } from 'fragments'
import { getMatter, getContact } from 'queries'
import withLoading from 'lib/withLoading'
import { renderTextarea, closeModalWindow } from 'utils'
import Button from 'components/Button'

const getMatters = gql`
  query myMatterReportMatter($after: String, $staffName: String) {
    matters(first: 15, after: $after, staffName: $staffName) {
      edges {
        cursor
        node {
          id
          name
          totalTimeValue
          totalTimeInvoiced
          wip
          billableStatusDisplay
          matterStatusDisplay
          daysOpen
          matterStatus
          budget
          description
          billingMethod
          client {
            id
            name
          }
          principal {
            id
          }
          manager {
            id
          }
          matterType {
            id
          }
          matterSubType {
            id
          }
          lastNote {
            id
            text
          }
        }
      }
      pageInfo {
        endCursor
        hasNextPage
      }
    }
  }
`

const NoteForm = ({
  matterId,
  contactId,
  mutate,
  matters,
  changeLoadStatus,
  isLoading,
  handleSubmit,
  dispatch,
  user,
}) => {
  const onSubmit = async (values) => {
    changeLoadStatus(true)
    if (matterId) {
      const response = await mutate({
        variables: {
          ...values,
          id: matterId,
          userId: user.id,
          dateTime: moment().format('YYYY-MM-DD'),
        },
        update: (store, { data: { createNote } }) => {
          if (matters) {
            const data = store.readQuery({
              query: getMatters,
              variables: { staffName: user.fullName },
            })
            data.matters.edges.find(({ node }) => node.id === matterId).node.lastNote =
              createNote.note
            store.writeQuery({
              query: getMatters,
              variables: { staffName: user.fullName },
              data,
            })
          } else {
            const data = store.readQuery({ query: getMatter, variables: { id: matterId } })
            data.matter.notes.unshift(createNote.note)
            store.writeQuery({
              query: getMatter,
              variables: { id: matterId },
              data,
            })
          }
        },
      })
      if (response.data.createNote.errors.length > 0) {
        console.log(response.data.createNote.errors)
      } else {
        dispatch(reset('notesForm'))
        closeModalWindow()
      }
    } else {
      const response = await mutate({
        variables: {
          ...values,
          id: contactId,
          userId: user.id,
          dateTime: moment().format('YYYY-MM-DD'),
        },
        update: (store, { data: { createNote } }) => {
          const data = store.readQuery({ query: getContact, variables: { id: contactId } })
          data.contact.notes.unshift(createNote.note)
          store.writeQuery({
            query: getContact,
            variables: { id: contactId },
            data,
          })
        },
      })
      if (response.data.createNote.errors.length > 0) {
        console.log(response.data.createNote.errors)
      } else {
        dispatch(reset('notesForm'))
        closeModalWindow()
      }
    }

    changeLoadStatus(false)
  }
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <Field name="id" component="input" type="hidden" />
      <Field
        style={{ minHeight: 200 }}
        component={renderTextarea}
        name="text"
        placeholder="Type a note..."
        id="text"
      />
      <Button
        title="Save"
        loading={isLoading}
        className="btn btn-success pull-right"
        type="submit"
        style={{ width: 120 }}
      />
    </form>
  )
}

const createNote = gql`
  mutation createNote($id: ID!, $text: String!, $userId: ID!, $dateTime: String!) {
    createNote(id: $id, text: $text, userId: $userId, dateTime: $dateTime) {
      errors
      note {
        ...Note
      }
    }
  }
  ${noteFragment}
`

export default reduxForm({
  form: 'notesForm',
  enableReinitialize: true,
})(graphql(createNote)(withLoading(connect()(NoteForm))))
