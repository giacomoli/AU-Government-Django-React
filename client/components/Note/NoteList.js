import React from 'react'

import Note from './Note'

export default ({ notes, matterId, contactId, user }) => {
  if (notes.length > 0) {
    return (
      <table id="mainTable" className="table table-bordered table-striped m-b-0">
        <thead>
          <tr>
            <th>Date time</th>
            <th width="70%">Note</th>
            <th>Staff member</th>
          </tr>
        </thead>
        <tbody>
          {notes.map(note => (
            <Note
              key={note.id}
              matterId={matterId}
              contactId={contactId}
              user={user}
              form={`note${note.id}`}
              note={note}
            />
          ))}
        </tbody>
        <style jsx global>{`
          .delete {
            width: 10px;
            text-align: center;
          }
        `}</style>
      </table>
    )
  }

  return <div>No notes</div>
}
