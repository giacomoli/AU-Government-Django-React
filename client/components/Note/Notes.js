import React from 'react'

import Modal from 'components/Modal'
import NoteForm from './NoteForm'
import NoteList from './NoteList'

export default ({ matterId, contactId, notes, user }) => [
  <Modal id="addNoteModal" title="Add a note" key={1}>
    <NoteForm matterId={matterId} contactId={contactId} user={user} />
  </Modal>,
  <div key={2} className="table-responsive">
    <div className="dt-buttons float-right col-xs-12">
      <button className="dt-button buttons-html5" data-toggle="modal" data-target="#addNoteModal">
        <span>Add</span>
      </button>
    </div>
    <NoteList notes={notes} user={user} matterId={matterId} contactId={contactId} />
    <style jsx>{`
      button {
        cursor: pointer;
      }
    `}</style>
  </div>,
]
