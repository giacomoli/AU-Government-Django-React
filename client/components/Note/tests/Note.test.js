import React from 'react'
import renderer from 'react-test-renderer'
import { ApolloProvider } from 'react-apollo';
import { reducer } from 'redux-form'
import { ApolloClient } from 'apollo-client';
import { createStore } from 'redux';

import Note from '../Note'

/* eslint-disable */
describe('Note', () => {

  const client = new ApolloClient({})
  const store = createStore(reducer)


  it('should render Note with some props', () => {
    const wrapper = renderer.create(
      <ApolloProvider client={client} store={store}>
        <Note
          note={{
            dateTime: '2018-10-02T14:30:00+00:00',
          }}
          form='note1'
        />
      </ApolloProvider>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should render Note with full props', () => {
    const wrapper = renderer.create(
      <ApolloProvider client={client} store={store}>
        <Note
          key={1}
          matterId={123123}
          contactId={12312}
          note={{
            dateTime: '2018-10-02T14:30:00+00:00',
            text: 'Test note',
            user: {
              fullName: 'Admin Test',
            },
          }}
          form="note1"
        />
      </ApolloProvider>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
