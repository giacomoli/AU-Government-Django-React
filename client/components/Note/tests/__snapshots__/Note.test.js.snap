// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note should render Note with full props 1`] = `
<tr
  className="jsx-2376464361"
>
  <td
    className="jsx-2376464361"
    style={
      Object {
        "textAlign": "center",
        "width": 200,
      }
    }
  >
    <div
      className="col"
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      Oct 2nd 2018
    </div>
  </td>
  <td
    className="jsx-2376464361"
  >
    <div
      className="jsx-2376464361 row"
    >
      <div
        className="col"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      >
        Test note
      </div>
    </div>
  </td>
  <td
    className="jsx-2376464361"
  >
    Admin Test
  </td>
</tr>
`;

exports[`Note should render Note with some props 1`] = `
<tr
  className="jsx-2376464361"
>
  <td
    className="jsx-2376464361"
    style={
      Object {
        "textAlign": "center",
        "width": 200,
      }
    }
  >
    <div
      className="col"
      onClick={[Function]}
      role="button"
      tabIndex={0}
    >
      Oct 2nd 2018
    </div>
  </td>
  <td
    className="jsx-2376464361"
  >
    <div
      className="jsx-2376464361 row"
    >
      <div
        className="col"
        onClick={[Function]}
        role="button"
        tabIndex={0}
      />
    </div>
  </td>
  <td
    className="jsx-2376464361"
  />
</tr>
`;
